### PR TITLE
Fix x.py install not working with relative prefix

### PR DESCRIPTION
**Pull request by emilyalbini**
The code powering `./x.py install` did not handle relative paths well: the installation script is executed inside a temporary directory, so all the relative paths specified in `config.toml` and in the `DESTDIR` environment variable were relative to that temporary directory. The original code fixed the problem for `config.toml` paths by canonicalizing the prefix, but breaking `DESTDIR`. https://github.com/rust-lang/rust/pull/80240 fixed the `DESTDIR` problem, but also regressed `config.toml` paths (#80683).

This PR refactors the installation code to generate paths that *in my understanding* are correct, adding comments in the meantime to explain what each step does. There was no documentation on why choices were made before, so my understanding could actually be wrong.

Regardless, executed `./x.py install` with various combinations of `config.toml` and `DESTDIR` paths, and everything seems to work according to my understanding. Still, I'd love if @vext01 and @yshui could test these changes.

r? @Mark-Simulacrum 
@rustbot modify labels: beta-nominated T-infra

### CODE REVIEW

**Comment by pietroalbini** [comment]
cc @cuviper as well, for a double-check from someone with distro experience. I *really* don't want to screw things here.

**Comment by cuviper** [comment]
The installer script appears to support `--destdir` -- should we just use that instead?

https://github.com/rust-lang/rust-installer/blob/5254dbfd25d5284728ab624dca1969d61427a0db/install-template.sh#L803

**Comment by cuviper** [comment]
Actually, I'm not sure that `--destdir` will prepend everything -- looks like it's only applied to the *default* paths.

**Comment by vext01** [comment]
Using your branch, I now see binaries, rlibs etc under the install prefix :)

File: src/bootstrap/install.rs [hunk 1]
```diff
@@ -26,74 +26,63 @@ fn install_sh(
...
+            }
         }
     }
-    ret
+
+    // The installation command is not executed from the current directory, but from a temporary
+    // directory. To prevent relative paths from breaking this converts relative paths to absolute
+    // paths. std::fs::canonicalize is not used as that requires the path to actually be present.
+    if path.is_relative() {
+        path = std::env::current_dir().expect("failed to get the current directory").join(path);
```
Comment (Mark-Simulacrum) [inline src/bootstrap/install.rs:81]: I think current_dir might be confusing here. I would expect it to be config.toml relative probably? I've never used install with relative paths though, and it seems fine to leave this up to future PRs to figure out. It's likely to not be entirely wrong at least.

**Pull request merged**